### PR TITLE
Blockquote refinements.

### PR DESCRIPTION
**Pull request by diegoreymendez**
Fixes [this issue](https://github.com/wordpress-mobile/WordPress-Editor-iOS/issues/597).

Heres [a video showing how it's working now](https://cloudup.com/cWNrHMx8MAd).

**Notes for testing:**
- Test in iOS 7 and 8.  These changes are big and we really want to avoid regressions.
- Select multiple lines or one line and press the blockquote button.
- Try to break blockquotes by editing the html by hand.  This isn't a big deal though, unless there's any important loss of data involved.

### CODE REVIEW

**Comment by bummytime** [comment]
@diegoreymendez I noticed that text is being removed when selecting both blockquoted and non-clockquoted text. Also looks like there are some rogue `<span>` tags making it into html mode:

https://cloudup.com/czmPCusTHjb

**Comment by bummytime** [comment]
@diegoreymendez were you ready for me to take another looksie at this?

**Comment by diegoreymendez** [comment]
@bummytime - Still on it... I'm having a hard time finding the right logic to fix the issue you found.  Want me to close the PR until it's ready again?

**Comment by bummytime** [comment]
@diegoreymendez Nah, I just appended "WIP" to the end of the PR title. Lemme know if you want/need any helps. :grinning:

**Comment by diegoreymendez** [comment]
@bummytime - Updated!  Finally :)

It still fixes the above bugs and some extra stuff.  Paragraph support is improved so it should be more difficult for the user to type outside of paragraphs - which in turn is a good thing to better stick to the web editor support.

I'm pretty sure we'll still find issues with blockquotes, especially when adding them inside other elements (such as lists).

My idea at this point would be to merge this in, if during your review we feel this is better than what we have, and open bugs for anything minor you may find.

Let me know your thoughts.

**Test 1:**
- Play with blockquotes, enable and disable at will.
- Make sure nothing breaks.  I'll open new issues for any minor bugs we find.
- Blockers and major issues will still be fixed in this PR.

**Test 2: **
- Try to write outside of paragraphs.  It can still be possible by editing the sources or pasting but it should be almost imposible to write outside paragraphs by typing in the visual editor.

**Test 3: **
- Make sure the issue reported at the top of this PR is fixed.

**Comment by diegoreymendez** [comment]
PS: @bummytime - most of the changes are in the project file, not sure why but I'll deal with those when merging. The PR is not that huge so don't worry. :)

**Comment by diegoreymendez** [comment]
PS2: @bummytime - also had to update some rangy files.  You should be able to skip those entirely.

**Comment by bummytime** [comment]
@diegoreymendez when running through this I noticed the issue I described [above](https://github.com/wordpress-mobile/WordPress-Editor-iOS/pull/622#issuecomment-91594282) still exists. Was this something you specifically fixed with the latest updates?

File: Assets/ZSSRichTextEditor.js [hunk 1]
```diff
@@ -346,6 +329,29 @@ ZSSEditor.getYCaretInfo = function() {
     return this.caretInfo;
 };
 
+/**
+ *  @brief      Whenever this method is called, a check will be performed on the caret position
+ *              to figure out if it needs to be wrapped in a paragraph node.
+ */
```
Comment (bummytime) [inline Assets/ZSSRichTextEditor.js]: Could we describe in a little more detail what kind of check we are performing in this function?

File: Assets/ZSSRichTextEditor.js [hunk 2]
```diff
@@ -633,6 +661,69 @@ ZSSEditor.quickLink = function() {
 	ZSSEditor.insertHTML(html_code);
 };
 
+// MARK: - Blockquotes
+
+ZSSEditor.toggleBlockquoteForSpecificChildNodes = function(parentNode, nodes) {
```
Comment (bummytime) [inline Assets/ZSSRichTextEditor.js]: Could you put a `@brief` and/or `@details` comment here?

File: Assets/ZSSRichTextEditor.js [hunk 3]
```diff
@@ -1875,6 +2163,24 @@ ZSSEditor.closerParentNode = function() {
     return parentNode;
 };
 
+ZSSEditor.closerParentElementRelativeToNode = function(node) {
```
Comment (bummytime) [inline Assets/ZSSRichTextEditor.js]: Is this function still being used?

**Comment by bummytime** [comment]
@diegoreymendez from the github issue noted at the top of the PR:

> we're adding an extra br or blank line of sorts at the end of blockquotes. See this post published from wp-admin and how it renders on the mobile editor

it looks like this is still happening (at least visually):

napkin 4 05-13-15 4 46 05 pm

Was this something that was going to be addressed in this PR?

**Comment by bummytime** [comment]
Really nice work @diegoreymendez. Let me know your thoughts on some of my feedback.

File: Assets/ZSSRichTextEditor.js (hunk 1 above)
Comment (diegoreymendez) [inline Assets/ZSSRichTextEditor.js]: Done!  And more!  Read below...

This was actually very enlightening and an awesome example of why documentation is important.  After reviewing the comments and seeing the code I realized something was wrong.  I ended up:

1. Moving the code to the field class.
2. Adding some controls that were missing from this method, and that where outside of it when they shouldn't have been.

While fixing this I also found and fixed a related issue with the title field, that was showing some strange characters when trying to delete or modify it.

File: Assets/ZSSRichTextEditor.js (hunk 2 above)
Comment (diegoreymendez) [inline Assets/ZSSRichTextEditor.js]: Done!

File: Assets/ZSSRichTextEditor.js (hunk 3 above)
Comment (diegoreymendez) [inline Assets/ZSSRichTextEditor.js]: No... there's a better alternative which is using the property `parentElement`.  I just removed this method.

**Comment by diegoreymendez** [comment]
@bummytime - Yes it was.  Although at this point I'd rather not tick that item as completed yet, and treat that issue in a different PR (to wrap this one up and start from scratch).  Does that sound good to you?

Btw, this review has been extremely positive for the code.  Very nice job!

**Comment by bummytime** [comment]
> Yes it was. Although at this point I'd rather not tick that item as completed yet, and treat that issue in a different PR (to wrap this one up and start from scratch). Does that sound good to you?

:+1:

**Comment by diegoreymendez** [comment]
I just fixed the exception you found in your tests and fixed the bug that was making some items disappear when enabling blockquotes.

**Comment by bummytime** [comment]
Nice @diegoreymendez ! :shipit:

**Comment by bummytime** [comment]
Hey @diegoreymendez I just noticed rangy 1.3 official just came out a few days ago. Looks like we are using 1.3 beta 2 here...might want to upgrade it.

**Comment by diegoreymendez** [comment]
Considering this is working as it is I'd prefer to split this new update into a new PR, if that sounds good to you.

**Comment by bummytime** [comment]
:+1:

**Comment by diegoreymendez** [comment]
Hey bummy, just double checking if this is ready to ship.  Let me know!

**Comment by bummytime** [comment]
@diegoreymendez yeps! :shipit:

**Pull request merged**